README.md

```diff
@@ -110,94 +103,66 @@ for more details.
+A HTTP request will be sent by a client to the [Server](#server).
+The `Server` will receive this request on the `data` event and
+is responsible to create a request object from this data.
+This object will be passed to the callback function.
+ 
+The request is an instance of [PSR-7 RequestInterface](https://github.com/php-fig/fig-standards/blob/master/accepted/PSR-7-http-message.md#32-psrhttpmessagerequestinterface).
+
+```php 
+$http = new Server($socket, function (RequestInterface $request, Response $response) {
+    $response->writeHead(200, array('Content-Type' => 'text/plain'));
+    $response->write("Request method: " . $request->getMethod() . "\n");
+    if ($request->getBody()->getSize() !== null) {
+        $response->write("Request body size: " . $request->getBody()->getSize() . "\n");
+    }
+    $response->end("The requested path is: " . $request->getUri()->getPath());
+});
+```
 
-It implements the `ReadableStreamInterface`.
+The `getBody()` method returns an instance of a [ReactPHP ReadableStreamInterface](https://github.com/reactphp/stream#readablestreaminterface)
+which implements the [PSR-7 StreamInterface](http://www.php-fig.org/psr/psr-7/#psrhttpmessagestreaminterface).
 
-Listen on the `data` event and the `end` event of the [Request](#request)
+Some functions of the `PSR-7 StreamInterface` are not used:
+`tell`, `eof`, `seek`, `rewind`, `write` and `read`.
+These methods SHOULD NOT be called, because these are not compatible
+with the `ReactPHP ReadableStreamInterface`.
+
+Listen on the `data` event and the `end` event of the body of the [Request](#request)
 to evaluate the data of the request body:
 
 ```php
-$http = new Server($socket, function (Request $request, Response $response) {
+$http = new Server($socket, function (RequestInterface $request, Response $response) {
     $contentLength = 0;
-    $request->on('data', function ($data) use (&$contentLength) {
+    $body = $request->getBody();
+    $body->on('data', function ($data) use (&$contentLength) {
         $contentLength += strlen($data);
     });
 
-    $request->on('end', function () use ($response, &$contentLength){
+    $body->on('end', function () use ($response, &$contentLength){
         $response->writeHead(200, array('Content-Type' => 'text/plain'));
         $response->end("The length of the submitted request body is: " . $contentLength);
     });
 
     // an error occures e.g. on invalid chunked encoded data or an unexpected 'end' event 
-    $request->on('error', function (\Exception $exception) use ($response, &$contentLength) {
+    $body->on('error', function (\Exception $exception) use ($response, &$contentLength) {
         $response->writeHead(400, array('Content-Type' => 'text/plain'));
         $response->end("An error occured while reading at length: " . $contentLength);
     });
 });
 ```
 
+If the request body is chunked-encoded, the data will be decoded and emitted on the data event.
+The `Transfer-Encoding` header will be removed.
```

Comment:
This appears a bit unclear, see also above.

Also note its position, how about referring to the above example first and then proceeding with these internals? (*telling a story*)

How about this:

```
The above example simply counts the number of bytes received in the request body.
This can be used as a skeleton for buffering or processing the request body.

If you only want to know the request body size, you can use its `getSize()` method.
This method returns the complete size of the request body as defined by the message boundaries.
This value may be `0` if the request message does not contain a request body (such as a simple `GET` request).
Note that this value may be `null` if the request body size is unknown in advance because the request message uses chunked transfer encoding.

[simple example from above here?]

The server automatically takes care of decoding chunked transfer encoding and will only emit the actual payload as data.
The `Transfer-Encoding` header will be removed.
```